Corrupted resource .cpp file when using QT5_ADD_RESOURCES and multithread compiling with CMake

longhuan2018 opened the pull request:
There are some problems with using QT5_ADD_RESOURCES to compile code under multithreading. For details, please refer to the link below.
https://stackoverflow.com/questions/23041356/corrupted-resource-cpp-file-when-using-qt5-add-resources-and-multithread-compil
This problem can be reproduced in the server and oauth2 projects. Because the file generated by QT5_ADD_RESOURCES was added to 2 projects at the same time to compile.
```
# add image resources
SET(QGIS_SERVER_IMAGERCCS ../../images/images.qrc)
QT5_ADD_RESOURCES(QGIS_SERVER_IMAGERCC_SRCS ${QGIS_SERVER_IMAGERCCS})

# add test resources, e.g. standard test font
SET(QGIS_SERVER_TESTRCCS ../../tests/testdata/testdata.qrc)
QT5_ADD_RESOURCES(QGIS_SERVER_TESTRCC_SRCS ${QGIS_SERVER_TESTRCCS})

ADD_EXECUTABLE(qgis_mapserv.fcgi qgis_map_serv.cpp ${QGIS_SERVER_TESTRCC_SRCS} ${QGIS_SERVER_IMAGERCC_SRCS})
ADD_EXECUTABLE(qgis_mapserver qgis_mapserver.cpp ${QGIS_SERVER_TESTRCC_SRCS} ${QGIS_SERVER_IMAGERCC_SRCS})
```
```
QT5_ADD_RESOURCES(OAUTH2_RCC_SRCS ${OAUTH2_RCCS})
############################################################
# Set up plugin targets

ADD_LIBRARY (oauth2authmethod MODULE
  ${OAUTH2_SRCS}
  ${OAUTH2_HDRS}
  ${OAUTH2_RCC_SRCS}
  ${OAUTH2_UIS_H}
)

# For unit testing
ADD_LIBRARY (oauth2authmethod_static STATIC
  ${OAUTH2_SRCS}
  ${OAUTH2_HDRS}
  ${OAUTH2_RCC_SRCS}
  ${OAUTH2_UIS_H}
)
```
There is also a description of QT5_ADD_RESOURCES in QT
```
Note: This is a low-level macro. See the CMake AUTORCC Documentation for a more convenient way to let Qt resource files be processed with . For embedding bigger resources, see qt5_add_big_resources.rcc
```
https://doc.qt.io/qt-5/qtcore-cmake-qt5-add-resources.html
https://cmake.org/cmake/help/latest/manual/cmake-qt.7.html#autorcc

longhuan2018 committed:
Use CMake's AUTOCC function instead of QT5_ADD_RESOURCES

longhuan2018 committed:
SET CMAKE_AUTORCC ON

longhuan2018 committed:
Update CMakeLists.txt

m-kuhn commented:
Good move to switch.
Would you be able to move IMAGE_RCCS to qgis_core at the same time? See https://github.com/qgis/QGIS/pull/33763#issuecomment-602157653?

longhuan2018 committed:
move IMAGE_RCCS to qgis_core

longhuan2018 commented:
> Good move to switch.
> Would you be able to move IMAGE_RCCS to qgis_core at the same time? See [#33763 (comment)](https://github.com/qgis/QGIS/pull/33763#issuecomment-602157653)?

I modified the following code according to your proposal, you see if it is appropriate to do so

m-kuhn commented:
Thanks.

I guess we can remove this now in most places. Since pretty much everything is linked to qgis_core already.

longhuan2018 commented:
> Thanks.
> 
> I guess we can remove this now in most places. Since pretty much everything is linked to qgis_core already.

OK, I will delete all the places where images.qrc is used

longhuan2018 committed:
remove all the places where images.qrc is used

m-kuhn reviewed (approved):
Looks good to me.

Is a backport of this required? If possible I'd like to skip a backport.

longhuan2018 committed:
fix test failed

QgsLayoutItemPicture :: loadPicture test result is incorrect due to successful loading of ": /images/composer/missing_image.svg"

m-kuhn commented on tests/src/app/testqgsapplayoutvaliditychecks.cpp:
```diff
@@ -222,13 +222,12 @@ void TestQgsLayoutValidityChecks::testPictureValidity()
   QList< QgsValidityCheckResult > res = check.runCheck( &context, &f );
-  QCOMPARE( res.size(), 1 );
-  QCOMPARE( res.at( 0 ).type, QgsValidityCheckResult::Warning );
+  QCOMPARE( res.size(), 0 );
```
@nyalldawson how does that look to you?

nyalldawson replied:
No, the original result was correct

longhuan2018 replied:
When we move IMAGE_RCCS to qgis_core, QString badFile (QStringLiteral (": /images/composer/missing_image.svg")) in QgsLayoutItemPicture :: loadPicture will return a valid value. This is the reason for the different results.
```
void QgsLayoutItemPicture::loadPicture( const QVariant &data )
{
  ...
  if ( mMode != FormatUnknown ) //make sure we start with a new QImage
  {
    recalculateSize();
  }
  else if ( mHasExpressionError || !mEvaluatedPath.isEmpty() )
  {
    //trying to load an invalid file or bad expression, show cross picture
    mMode = FormatSVG;
    mIsMissingImage = true;
    QString badFile( QStringLiteral( ":/images/composer/missing_image.svg" ) );
    mSVG.load( badFile );
    if ( mSVG.isValid() )
    {
      mMode = FormatSVG;
      QRect viewBox = mSVG.viewBox(); //take width/height ratio from view box instead of default size
      mDefaultSvgSize.setWidth( viewBox.width() );
      mDefaultSvgSize.setHeight( viewBox.height() );
      recalculateSize();
    }
  }
}
```

nyalldawson replied:
Thanks -- you'll need https://github.com/qgis/QGIS/pull/35864 to fix this

longhuan2018 committed:
Merge pull request #9 from qgis/master

merge from qgis

longhuan2018 committed:
Revert test src changes

longhuan2018 commented:
@nyalldawson  
I merged the latest changes into this PR, and reverted the changes in the test code. 
But continuous-integration / travis-ci / pr failed, I do n’t know how to solve it.
Can you help me?

m-kuhn commented:
Have a look at the cdash results to get a better impression:

https://cdash.orfeo-toolbox.org/viewTest.php?onlyfailed&buildid=39165

nyalldawson commented:
Superseded by #35961 35969
